src/api/baggage.rs

```diff
@@ -150,46 +218,69 @@ pub trait BaggageExt {
     ///     Some(&Value::String("my-value".to_string())),
     /// )
     /// ```
-    fn with_baggage<T: IntoIterator<Item = KeyValue>>(&self, baggage: T) -> Self;
+    fn with_baggage<T: IntoIterator<Item = I>>(&self, baggage: T) -> Self;
 
-    /// Returns a clone of the given context with the included name / value pairs.
+    /// Returns a clone of the current context with the included name / value pairs.
     ///
     /// # Examples
     ///
     /// ```
-    /// use opentelemetry::api::{BaggageExt, Context, KeyValue, Value};
+    /// use opentelemetry::api::{BaggageExt, Context, KeyValue, Value, AddBaggage};
     ///
-    /// let cx = Context::current().with_cleared_baggage();
+    /// let cx = Context::current_with_baggage(vec![KeyValue::new("my-name", "my-value")]);
     ///
-    /// assert_eq!(cx.baggage().len(), 0);
+    /// assert_eq!(
+    ///     cx.baggage().get("my-name"),
+    ///     Some(&Value::String("my-value".to_string())),
+    /// )
     /// ```
-    fn with_cleared_baggage(&self) -> Self;
-
-    /// Returns a reference to this context's baggage, or the default
-    /// empty baggage if none has been set.
-    fn baggage(&self) -> &Baggage;
+    fn current_with_baggage<T: IntoIterator<Item = I>>(baggage: T) -> Self;
 }
+
 impl BaggageExt for Context {
-    fn current_with_baggage<T: IntoIterator<Item = KeyValue>>(kvs: T) -> Self {
-        Context::current().with_baggage(kvs)
+    fn with_cleared_baggage(&self) -> Self {
+        self.with_value(Baggage::new())
     }
 
-    fn with_baggage<T: IntoIterator<Item = KeyValue>>(&self, kvs: T) -> Self {
+    fn baggage(&self) -> &Baggage {
+        self.get::<Baggage>().unwrap_or_else(|| &DEFAULT_BAGGAGE)
+    }
+}
+
+impl AddBaggage<KeyValueMetadata> for Context {
+    fn with_baggage<T: IntoIterator<Item = KeyValueMetadata>>(&self, kvm: T) -> Self {
         let merged: Baggage = self
             .baggage()
             .iter()
-            .map(|(key, value)| KeyValue::new(key.clone(), value.clone()))
-            .chain(kvs.into_iter())
+            .map(|(key, (value, metadata))| {
```

Comment:
I don't particularly like the nested tuple from the iterator here. From the perspective or `KeyValue` it makes sense, because it maps a key to a (value, metadata) pair. But using this here feels weird. I may prefer this to be a flat tuple with 3 items (key, value, metadata). But that might just be me and I'm not super sure about this, either.

Reply:
I think have `key, (value, metadata)` give some clearity around which part is the key and which part is the value. A ternary `a,b,c` could be group in either `(a,b),c` or `a, (b,c)`. But it should be a easy change so we can also make it we'd like to